INTEGRATION.md

```diff
@@ -0,0 +1,152 @@
+1. a gateway with a response below 50ms has been found
+2. gateways have been probed for over 2 seconds (in which case it just stops and uses the fastest one).
+
+The result of this probe will be stored in `$CONFIG/ipfs/racing_gateway_response` as a single line being the gateway that responded fastest. In subsequent racing gateway requests this file will be read and used as starting point. If this gateway still responds within 50ms then no other gateways will be probed.
+
+\*** `dweb.link`, see the below `IPFS_FALLBACK_GATEWAY` for details.
+
+#### Environment variables
+
+The decision tree is influences by a couple environement variables.
+
+**`IPFS_RACING_GATEWAY`** When this environemnt variable isn't found (the default), racing gateways should be attempted upon reaching this point in the flow. When the environment variable exists it's value should be used. If it's value is `1` (or `true`) then racing gateways should be attempted. If this value is `0` (or `false`) then racing gateways are off. The control flow proceeds in the `no` branch.
+
+**`IPFS_FALLBACK_GATEWAY`** When this variable doesn't exist `dweb.link` will be used. When the variable does exist it's value will be used instead.
+
+### Gateway from command argument
+
+**This feature is optional and only for applications integrating IPFS support.**
+
+An application can opt to support a command line option to provide a gateway. If a user does provide this option then it should overrule any other gateway detection and be used as the gateway of choice. If implemented, it's recommended to go for either a `--gateway` or `--ipfs-gateway` argument. It depends very much on the application itself as to which option is most sensible.
+
+An example implementation that is doing this is ffmpeg with the ffplay utility. It allows the `-gateway` argument which by default is empty but can be set like: `-gateway http://127.0.0.1:8080` and would then be used to handle `ipfs://<cid>` or `ipns://<cid>`.
+
+### Gateway from IPFS_GATEWAY environment variable
+
+When there is no command line argument, the `IPFS_GATEWAY` environment is next. If it contains a value, it will be used as gateway.
+
+### Gateway file
+
+**This feature is only for IPFS implementers, not for IPFS application integrations.**
+
+When the implementation provides a gateway (and it's not disabled through other means) then it should make that known to other local applications. The gateway file serves this purpose. It's a file with only 1 single line containing the full http URL to your gateway. For example, it could contain the line: "http://localhost:8080".
+
+The file conditions:
+    1. is named "gateway"
+    2. **only** exists when your implementation actually starts a gateway
+    3. is removed when your implementation shuts down
+
+For historical and compatibility readons, this file shall be placed in:
+$HOME/.ipfs/ thus the resulting end path is going to be $HOME/.ipfs/gateway
```

Comment:
We need to clean this up – as-is, it won't work with Iroh.  cc @b5 

My take: 
- Specification should allow for running Iroh and Kubo on the same PC. 
- Canonical location should be implementation-agnostic, namely `$XDG_CONFIG_HOME/ipfs/gateway`.
Kubo should switch to that.
  - For backward-compatibility, Kubo would  create symlink from `$IPFS_PATH/gateway` to `$XDG_CONFIG_HOME/ipfs/gateway`